common/flatpak-context.c

```diff
@@ -888,6 +995,27 @@ flatpak_context_merge (FlatpakContext *context,
   while (g_hash_table_iter_next (&iter, &key, &value))
     g_hash_table_insert (context->persistent, g_strdup (key), value);
 
+  /* We first handle home and host with the RESET flag, as they override
+   * other keys than themselves from the parent */
+  if (other->fs_host_flags & FLATPAK_FILESYSTEM_FLAGS_RESET)
+    {
+      g_hash_table_remove_all (context->filesystems);
+    }
+  else if (other->fs_home_flags & FLATPAK_FILESYSTEM_FLAGS_RESET)
+    {
+      g_hash_table_iter_init (&iter, context->filesystems);
+
+      while (g_hash_table_iter_next (&iter, &key, &value))
+        {
+          if (flatpak_filesystem_key_in_home (key))
+            g_hash_table_iter_remove (&iter);
+        }
+    }
+
+  context->fs_home_flags |= other->fs_home_flags;
+  context->fs_host_flags |= other->fs_host_flags;
```

Comment:
This doesn't seem right. If the "upper" layer (`other` here) has "host" (not host:reset) but the lower layer has host:reset, then the result should be "host", but the |= will never drop the reset flag.

I think this needs to be
 if ("home" in other.filesystems)
   context.fs_home_fags = other.fs_home_flags

Reply:
> This doesn't seem right. If the "upper" layer (other here) has "host" (not host:reset) but the lower layer has host:reset, then the result should be "host", but the |= will never drop the reset flag.

Perhaps, let's reason through it.

By the time we get here, we have already decided whether to inherit filesystems from `context` or reset it to a blank slate, so the flags only matter for subsequent merges.

In practice, we start from a low-level context and successively merge higher-level contexts into it:

* lowest = Context()
* middle = ... (e.g. from overrides)
* highest = ... (e.g. from CLI)
* lowest.merge(other=middle)
* lowest.merge(other=highest)

so I don't think the `context` (lower level) flags ever actually matter in practice: it's the `other` (higher level) flags that affect our behaviour. So, right now, this is purely academic. However, if we switched to this pattern in future, or if there's an instance of it that I hadn't spotted:

* lowest = Context()
* middle = ...
* highest = ...
* middle.merge(other=highest)
* lowest.merge(other=middle)

then it starts to matter, because `lowest.merge(middle)` will respect the `middle->fs_host_flags`. (Or `home`, they're interchangeable).

For this to matter, we would need a stack like this:

* lowest (e.g. manifest): `--filesystem=/one`
* middle (e.g. overrides): `--nofilesystem=host:reset --filesystem=/two`
* highest (e.g. CLI): `--nofilesystem=host --filesystem=/three`

In either my interpretation or yours, the order of evaluation that we really use in practice would give us:

* `lowest` is initially `{fs_host_flags: NONE, filesystems={/one: RW}}`
* `lowest.merge(middle)` yields `{fs_host_flags: RESET, filesystems={/two: RW}}`
* `lowest.merge(highest)` yields `{fs_host_flags: ???, filesystems={/two: RW, /three: RW}}`

and the only difference between our interpretations is whether the flags end up being `NONE` (your interpretation) or `RESET` (my interpretation).

However, if we have the other order of evaluation somewhere, then that changes things. If we take my interpretation with `|=`, then we get:

* `middle.merge(other=highest)` will result in `middle` being `{fs_host_flags: RESET, filesystems={host: NONE, /two: RW, /three: RW}}`
* `lowest.merge(other=middle)` results in `lowest` being a copy of `middle`

but in your interpretation with conditional assignment:

* `middle.merge(other=highest)` will result in `middle` being `{fs_host_flags: NONE, filesystems={host: NONE, /two: RW, /three: RW}}`
* `lowest.merge(other=middle)` results in `lowest` being `{fs_host_flags: NONE, filesystems={host: NONE, /one: RW, /two: RW, /three: RW}}`

I think that's dangerous: reversing the order in which we do the stacked merges has resulted in `/one` being exposed unintentionally. So I think we do want my interpretation here, to get a more conservative result in the face of possible future implementation changes.

Reply:
I've gone backwards and forwards on whether the `RESET` flags should be stored in the `FlatpakContext` as a pair of flags words, or a pair of booleans.

If we want more flags in future, we can't really predict whether we will want them to have these "sticky" semantics or not, which might be an argument for storing them as a pair of booleans. That would mean that, if we add another flag in future, we are forced to think about how *that* flag needs to behave during merges, independent of how the `RESET` flag behaves during merges.

Reply:
Actually, I'm not sure that is right either.

If the lower layer has !host:remove, and the upper layer has `host`, then the merged should have host in filenames, and 0 flag.

However, if the lowerlayer has `!host:remove` and the upper layer has `!host`, then I agree with your reasoning above. For example, this would happen if you combine global and app overrides, one has `!host:remove` and the other has `!host´. In this example I would thing the expected behavior was the complete override.

So:
if home in other.filesystem:
   if other.filesystem[home] == NONE:
     context.fs_home_flags |= other.fs_home_flags
  else:
    context.fs_home_flags = 0

Does that make sense?

Reply:
> If the lower layer has !host:remove, and the upper layer has host, then the merged should have host in filenames, and 0 flag.

Hmm. That would mean that if we have this corner-case:

* manifest: --filesystem=/one
* system overrides: --nofilesystem=host:reset --filesystem=/two
* user overrides: --filesystem=host --filesystem=/three
* CLI: --nofilesystem=host --filesystem=/four

then the result depends on the order of evaluation. If we do `lowest.merge(sys_overrides).merge(user_overrides).merge(cli)` like we do now, then `/one` would end up being disabled by the `RESET`, and we would only share `/two`, `/three` and `/four`. But if we did `user_overrides.merge(cli); sys_overrides.merge(user_overrides); lowest.merge(sys_overrides)` then we would lose track of the `RESET` flag, and `/one` would end up enabled. It seems bad for order of evaluation to have an effect like that?

Reply:
This does lead to the interesting situation that merging can result in a context that has the `RESET` flag set, *and* has `host` access greater than `NONE`.

Reply:
Actually we're not merging overrides like that, but:

```
app = from metadata
all_overrides = system_override.merge(system_app_override).merge(user_override).merge(use_app_override)
cli = from cli

app.merge(all_overrides).merge(cli).
```
See flatpak_deploy_get_overrides().

So, yes, it does seems like a host with reset set is possible...

Reply:
I guess the question is, should we care about that case, or should we change how we handle multple overrides?

Reply:
Maybe using flags is a bad idea, and we should convert host:reset on the CLI to host-reset in the internal APIs, then the two would be completely separate with little special casing needed (other than when converting back to metadata/args.

In that case we naturally handle the !host-reset + host case.

Reply:
I added some test coverage for these corner-cases in `test-context.c`.

Reply:
I have kept the semantics I wrote, rather than taking your suggestion, because it doesn't matter most of the time and I think my semantics give us slightly better results in corner-cases.